### PR TITLE
JAMES-2294 WebAdmin should allow mail reprocessing.

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by chibenwa** [comment]
test this please (take advantage of available runners)

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java [hunk 1]
```diff
@@ -271,6 +283,100 @@ public void defineDeleteAll() {
...
+            value = "If present, modifies the state property of the mail to allow their processing by a specific mail container processor.")
+    })
+    @ApiResponses(value = {
+        @ApiResponse(code = HttpStatus.CREATED_201, message = "Task is created", response = TaskIdDto.class),
+        @ApiResponse(code = HttpStatus.INTERNAL_SERVER_ERROR_500, message = "Internal server error - Something went bad on the server side."),
+        @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "Bad request - unknown action")
+    })
+    public void defineReprocessAll() {
+        service.patch(MAIL_REPOSITORIES + "/:encodedUrl/mails", (request, response) -> {
+            String url = decodedRepositoryUrl(request);
```
Comment (aduprat) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java]: Can you extract this part?

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java [hunk 2]
```diff
@@ -271,6 +283,100 @@ public void defineDeleteAll() {
...
+            value = "If present, modifies the state property of the mail to allow their processing by a specific mail container processor.")
+    })
+    @ApiResponses(value = {
+        @ApiResponse(code = HttpStatus.CREATED_201, message = "Task is created", response = TaskIdDto.class),
+        @ApiResponse(code = HttpStatus.INTERNAL_SERVER_ERROR_500, message = "Internal server error - Something went bad on the server side."),
+        @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "Bad request - unknown action")
+    })
+    public void defineReprocessOne() {
+        service.patch(MAIL_REPOSITORIES + "/:encodedUrl/mails/:key", (request, response) -> {
+            String url = decodedRepositoryUrl(request);
```
Comment (aduprat) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java]: idem

File: server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/dto/TaskIdDto.java [hunk 3]
```diff
@@ -33,7 +33,7 @@
 
     public static TaskIdDto respond(Response response, TaskId taskId) {
         response.status(HttpStatus.CREATED_201);
-        response.header(LOCATION.asString(), TasksRoutes.BASE + "/" + taskId.toString());
+        response.header(LOCATION.asString(), TasksRoutes.BASE + "/" + taskId.getValue());
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/dto/TaskIdDto.java]: does it fix anything ? I don't see any test change. Or maybe toString returns the same thing than getValue ?

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/ReprocessingAllTask.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    private final AdditionalInformation additionalInformation;
+
+    public ReprocessingAllTask(ReprocessingService reprocessingService, MailRepositoryStoreService mailRepositoryStoreService,
+                               String repositoryUrl, String targetQueue, Optional<String> targetProcessor) {
+        this.reprocessingService = reprocessingService;
+        this.repositoryUrl = repositoryUrl;
+        this.targetQueue = targetQueue;
+        this.targetProcessor = targetProcessor;
+        this.additionalInformation = new AdditionalInformation(
+            Throwing.supplier(() -> mailRepositoryStoreService.size(repositoryUrl).orElse(0L)));
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/ReprocessingAllTask.java]: it looks like a bad idea. Having a DTO datastructure calling a service is dangerous IMO. We should rather update the status with a listener if needed.

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/ReprocessingOneTask.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    private final AdditionalInformation additionalInformation;
+
+    public ReprocessingOneTask(ReprocessingService reprocessingService,
+                               String repositoryUrl, String targetQueue, String mailKey, Optional<String> targetProcessor) {
+        this.reprocessingService = reprocessingService;
+        this.repositoryUrl = repositoryUrl;
+        this.targetQueue = targetQueue;
+        this.mailKey = mailKey;
+        this.targetProcessor = targetProcessor;
+        this.additionalInformation = new AdditionalInformation();
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/ReprocessingOneTask.java]: I don't really like the non-static class, it's really unusual

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java [hunk 6]
```diff
@@ -576,4 +591,583 @@ public void deleteShouldReturnNotFoundWhenNoMailRepository() throws Exception {
...
+        with()
+            .basePath(TasksRoutes.BASE)
+            .get(taskId + "/await");
+
+        assertThat(mailRepository.size())
+            .isEqualTo(2);
+    }
+
+    @Test
+    public void reprocessingOneTaskShouldFailWhenUnknownMailKey() throws Exception {
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: you have a mix of reprocessingOne and reprocessingOneTask in this test file. Can you keep reprocessingOne everywhere please?

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/ReprocessingService.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    private void reprocess(MailRepository repository, MailQueue mailQueue, String key, Optional<String> targetProcessor) throws MessagingException {
+        Mail mail = repository.retrieve(key);
+        targetProcessor.ifPresent(mail::setState);
+        mailQueue.enQueue(mail);
+        repository.remove(key);
+    }
+
+    private MailQueue getMailQueue(String targetQueue) {
+        return mailQueueFactory.getQueue(targetQueue)
+            .orElseThrow(() -> ErrorResponder.builder()
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/ReprocessingService.java]: it's the wrong layer to talk HTTP.

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java [hunk 8]
```diff
@@ -271,6 +283,100 @@ public void defineDeleteAll() {
...
+    @Path("/{encodedUrl}/mails")
+    @ApiOperation(value = "Reprocessing all mails in that mailRepository")
+    @ApiImplicitParams({
+        @ApiImplicitParam(
+            required = false,
+            name = "queue",
+            paramType = "query parameter",
+            dataType = "String",
+            defaultValue = "spool",
+            example = "?offset=outgoing",
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java]: offset ?

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java [hunk 9]
```diff
@@ -271,6 +283,100 @@ public void defineDeleteAll() {
...
+    })
+    @ApiResponses(value = {
+        @ApiResponse(code = HttpStatus.CREATED_201, message = "Task is created", response = TaskIdDto.class),
+        @ApiResponse(code = HttpStatus.INTERNAL_SERVER_ERROR_500, message = "Internal server error - Something went bad on the server side."),
+        @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "Bad request - unknown action")
+    })
+    public void defineReprocessAll() {
+        service.patch(MAIL_REPOSITORIES + "/:encodedUrl/mails", (request, response) -> {
+            String url = decodedRepositoryUrl(request);
+            enforceActionParameter(request);
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java]: there is no action parameter document, right ?

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java [hunk 10]
```diff
@@ -271,6 +283,100 @@ public void defineDeleteAll() {
...
+    @Path("/{encodedUrl}/mails/{key}")
+    @ApiOperation(value = "Reprocessing a single mail in that mailRepository")
+    @ApiImplicitParams({
+        @ApiImplicitParam(
+            required = false,
+            name = "queue",
+            paramType = "query parameter",
+            dataType = "String",
+            defaultValue = "spool",
+            example = "?offset=outgoing",
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java]: offset here too

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java [hunk 11]
```diff
@@ -271,6 +283,100 @@ public void defineDeleteAll() {
...
+    @ApiResponses(value = {
+        @ApiResponse(code = HttpStatus.CREATED_201, message = "Task is created", response = TaskIdDto.class),
+        @ApiResponse(code = HttpStatus.INTERNAL_SERVER_ERROR_500, message = "Internal server error - Something went bad on the server side."),
+        @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "Bad request - unknown action")
+    })
+    public void defineReprocessOne() {
+        service.patch(MAIL_REPOSITORIES + "/:encodedUrl/mails/:key", (request, response) -> {
+            String url = decodedRepositoryUrl(request);
+            String key = request.params("key");
+            enforceActionParameter(request);
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java]: same thing

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java [hunk 12]
```diff
@@ -576,4 +591,583 @@ public void deleteShouldReturnNotFoundWhenNoMailRepository() throws Exception {
...
+            .name(name1)
+            .build());
+        mailRepository.store(FakeMail.builder()
+            .name(name2)
+            .build());
+
+        String transport = "transport";
+        String taskId = with()
+            .patch(URL_ESCAPED_MY_REPO + "/mails?action=reprocess" +
+                "&queue=" + CUSTOM_QUEUE +
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: you should put params with `param()` method

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java (hunk 6 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: Sorry, probably a mistake...

File: server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/dto/TaskIdDto.java (hunk 3 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/dto/TaskIdDto.java]: Yes it fixes something. Gonna write a test.

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java (hunk 8 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java]: More probably `?queue=outgoing`

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java (hunk 9 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java]: Right... Rush you know...

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java (hunk 10 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java]: ... Gonna fix

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/ReprocessingOneTask.java (hunk 5 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/ReprocessingOneTask.java]: Wanted to try, it makes spare some fiew fields, however if you prefer, I gonna make it static.

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java (hunk 12 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: Gonna do it in a separated commit.

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/ReprocessingAllTask.java (hunk 4 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/ReprocessingAllTask.java]: Gonna get it a try, but I smell that it gonna be `diff hungry`

Comment (mbaechler) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/ReprocessingAllTask.java]: Do we need that information right now ? Maybe we can skip that field for now.

**Comment by aduprat** [comment]
```
[INFO] --- maven-checkstyle-plugin:2.17:check (default) @ james-server-webadmin-mailrepository ---
[INFO] There are 3 errors reported by Checkstyle 8.5 with checkstyle.xml ruleset.
[ERROR] src/main/java/org/apache/james/webadmin/service/ReprocessingService.java:[33,8] (imports) UnusedImports: Unused import - org.apache.james.webadmin.utils.ErrorResponder.
[ERROR] src/main/java/org/apache/james/webadmin/service/ReprocessingService.java:[35,8] (imports) UnusedImports: Unused import - org.eclipse.jetty.http.HttpStatus.
[ERROR] src/main/java/org/apache/james/webadmin/service/ReprocessingAllMailsTask.java:[32,8] (imports) UnusedImports: Unused import - com.github.fge.lambdas.Throwing.
[INFO] ------------------------------------------------------------------------
```
I will take care of those errors.

**Comment by aduprat** [comment]
> I will take care of those errors.
done

**Comment by chibenwa** [comment]
:+1:

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java [hunk 13]
```diff
@@ -217,8 +217,11 @@ public void listingKeysShouldApplyLimitAndOffset() throws Exception {
             .name("name3")
             .build());
 
-        when()
-            .get(MY_REPO_MAILS + "?offset=1&limit=1")
+        given()
+            .param("limit", "1")
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: ok, I was wrong, it's queryParam, not param.
You may wonder if you really test what you think

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToRepositoryTest.java [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+
+        awaitOneMinute.until(() -> probe.getRepositoryMailCount(CUSTOM_REPOSITORY) == 2);
+        String key = probe.listMailKeys(CUSTOM_REPOSITORY).get(0);
+
+        with()
+            .spec(webAdminAPI)
+            .patch(MailRepositoriesRoutes.MAIL_REPOSITORIES
+                + "/" + URLEncoder.encode(CUSTOM_REPOSITORY, StandardCharsets.UTF_8.displayName())
+                + "/mails/" + key
+                + "?processor=" + ProcessorConfiguration.STATE_TRANSPORT
```
Comment (mbaechler) [inline server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToRepositoryTest.java]: queryParam ?

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToRepositoryTest.java [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+            .awaitSent(awaitOneMinute);
+
+        awaitOneMinute.until(() -> probe.getRepositoryMailCount(CUSTOM_REPOSITORY) == 2);
+
+        with()
+            .spec(webAdminAPI)
+            .patch(MailRepositoriesRoutes.MAIL_REPOSITORIES
+                + "/" + URLEncoder.encode(CUSTOM_REPOSITORY, StandardCharsets.UTF_8.displayName())
+                + "/mails"
+                + "?processor=" + ProcessorConfiguration.STATE_TRANSPORT
```
Comment (mbaechler) [inline server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToRepositoryTest.java]: queryparam ?

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToRepositoryTest.java [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+    @After
+    public void tearDown() {
+        jamesServer.shutdown();
+    }
+
+    @Test
+    public void incomingShouldBeStoredInProcessorByDefault() throws Exception {
+        messageSender.connect(LOCALHOST_IP, SMTP_PORT)
+            .sendMessage(RECIPIENT, RECIPIENT)
+            .awaitSent(awaitOneMinute)
```
Comment (mbaechler) [inline server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToRepositoryTest.java]: it should probably be called awaitAtMostOneMinute, right ?

File: server/protocols/webadmin/webadmin-core/src/test/java/org/apache/james/webadmin/dto/TaskIdDtoTest.java [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+import org.eclipse.jetty.http.HttpStatus;
+import org.junit.Test;
+
+import spark.Response;
+
+public class TaskIdDtoTest {
+    private static final String UID_VALUE = "ce5316cb-c924-40eb-9ca0-c5828e276297";
+
+    @Test
+    public void test() {
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-core/src/test/java/org/apache/james/webadmin/dto/TaskIdDtoTest.java]: not a good test name

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java (hunk 13 above)
Comment (rouazana) [reply server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: I checked and both seems to do the same (even checked with wireshark). What's the difference?

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToRepositoryTest.java (hunk 16 above)
Comment (rouazana) [reply server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToRepositoryTest.java]: not related but ok

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**